File: headroom/rollout.py
"""Release-channel and feature-flag policy for Headroom.

The rollout layer is intentionally small: it answers whether a named feature is
eligible in the current release channel and whether it is enabled by default or
explicit request. Composition roots should depend on this module instead of
reading feature env vars directly.
"""

from __future__ import annotations

import os
from collections.abc import Mapping
from dataclasses import dataclass
from enum import Enum

_TRUE_VALUES = {"1", "true", "yes", "on", "enabled"}
_FALSE_VALUES = {"0", "false", "no", "off", "disabled"}


class ReleaseChannel(str, Enum):
    """Ordered stability channels used to bake features before stable release."""

    STABLE = "stable"
    BETA = "beta"
    CANARY = "canary"
    DEV = "dev"

    @classmethod
    def parse(cls, value: str | None) -> ReleaseChannel:
        if not value:
            return cls.STABLE
        normalized = value.strip().lower().replace("-", "_")
        aliases = {
            "prod": cls.STABLE,
            "production": cls.STABLE,
            "preview": cls.BETA,
            "nightly": cls.CANARY,
            "development": cls.DEV,
        }
        if normalized in aliases:
            return aliases[normalized]
        try:
            return cls(normalized)
        except ValueError:
            return cls.STABLE

    @property
    def order(self) -> int:
        return {
            ReleaseChannel.STABLE: 0,
            ReleaseChannel.BETA: 1,
            ReleaseChannel.CANARY: 2,
            ReleaseChannel.DEV: 3,
        }[self]

    def allows(self, required: ReleaseChannel) -> bool:
        return self.order >= required.order


@dataclass(frozen=True)
class FeatureSpec:
    """Static policy for one rollout-managed feature."""

    name: str
    available_in: ReleaseChannel
    default_enabled_in: ReleaseChannel | None = None
    legacy_env: tuple[str, ...] = ()
    description: str = ""

    def default_enabled(self, channel: ReleaseChannel) -> bool:
        return self.default_enabled_in is not None and channel.allows(self.default_enabled_in)


FEATURES: dict[str, FeatureSpec] = {
    "tool_result_interceptors": FeatureSpec(
        name="tool_result_interceptors",
        available_in=ReleaseChannel.CANARY,
        default_enabled_in=None,
        legacy_env=("HEADROOM_INTERCEPT_ENABLED",),
        description="AST-aware Read/tool-result interceptors used before compression.",
    ),
    "proxy_output_shaper": FeatureSpec(
        name="proxy_output_shaper",
        available_in=ReleaseChannel.BETA,
        default_enabled_in=None,
        legacy_env=("HEADROOM_OUTPUT_SHAPER",),
        description="Proxy output-shaping path for response-side experiments.",
    ),
    "read_maturation": FeatureSpec(
        name="read_maturation",
        available_in=ReleaseChannel.BETA,
        default_enabled_in=None,
        legacy_env=("HEADROOM_READ_MATURATION",),
        description="Hold-back Read maturation before provider cache entry.",
    ),
}


def _split_names(raw: str | None) -> set[str]:
    if not raw:
        return set()
    return {
        part.strip().lower().replace("-", "_")
        for part in raw.replace(";", ",").split(",")
        if part.strip()
    }


def _truthy(value: str | None) -> bool:
    return bool(value and value.strip().lower() in _TRUE_VALUES)


def _falsey(value: str | None) -> bool:
    return bool(value and value.strip().lower() in _FALSE_VALUES)


@dataclass(frozen=True)
class Rollout:
    """Resolved runtime rollout state."""

    channel: ReleaseChannel
    enabled: frozenset[str]
    disabled: frozenset[str]
    unsafe_allow_unstable: bool = False

    @classmethod
    def from_env(cls, environ: Mapping[str, str] | None = None) -> Rollout:
        env = os.environ if environ is None else environ
        return cls(
            channel=ReleaseChannel.parse(env.get("HEADROOM_RELEASE_CHANNEL")),
            enabled=frozenset(_split_names(env.get("HEADROOM_FEATURES"))),
            disabled=frozenset(_split_names(env.get("HEADROOM_DISABLE_FEATURES"))),
            unsafe_allow_unstable=_truthy(env.get("HEADROOM_UNSAFE_ALLOW_UNSTABLE_FEATURES")),
        )

    def is_available(self, feature: str) -> bool:
        spec = FEATURES[feature]
        return self.channel.allows(spec.available_in) or self.unsafe_allow_unstable

    def is_enabled(
        self,
        feature: str,
        *,
        explicit: bool = False,
        environ: Mapping[str, str] | None = None,
    ) -> bool:
        """Return whether a feature may run in this process.

        `explicit=True` represents a typed configuration or CLI option. Legacy
        per-feature env vars are treated as explicit requests for migration, but
        they still obey the channel gate.
        """
        spec = FEATURES[feature]
        if feature in self.disabled:
            return False
        if not self.is_available(feature):
            return False
        if feature in self.enabled:
            return True
        if explicit:
            return True

        env = os.environ if environ is None else environ
        for name in spec.legacy_env:
            raw = env.get(name)
            if _falsey(raw):
                return False
            if _truthy(raw):
                return True

        return spec.default_enabled(self.channel)


def current_rollout(environ: Mapping[str, str] | None = None) -> Rollout:
    """Resolve rollout state from the process environment."""

    return Rollout.from_env(environ)


def feature_enabled(
    feature: str,
    *,
    explicit: bool = False,
    environ: Mapping[str, str] | None = None,
) -> bool:
    """Convenience wrapper for composition roots."""

    return current_rollout(environ).is_enabled(feature, explicit=explicit, environ=environ)

File: crates/headroom-core/src/rollout.rs
//! Release-channel and feature-flag policy shared by Rust components.
//!
//! Runtime feature gates should be resolved in composition roots. Lower-level
//! modules receive concrete collaborators or config values, not ad-hoc env
//! reads, so experiments cannot spread conditional logic through hot paths.

use std::collections::BTreeSet;
use std::str::FromStr;

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
pub enum ReleaseChannel {
    #[default]
    Stable,
    Beta,
    Canary,
    Dev,
}

impl ReleaseChannel {
    pub fn as_str(self) -> &'static str {
        match self {
            ReleaseChannel::Stable => "stable",
            ReleaseChannel::Beta => "beta",
            ReleaseChannel::Canary => "canary",
            ReleaseChannel::Dev => "dev",
        }
    }

    pub fn allows(self, required: ReleaseChannel) -> bool {
        self >= required
    }
}

impl FromStr for ReleaseChannel {
    type Err = ();

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value.trim().to_ascii_lowercase().replace('-', "_").as_str() {
            "" | "stable" | "prod" | "production" => Ok(Self::Stable),
            "beta" | "preview" => Ok(Self::Beta),
            "canary" | "nightly" => Ok(Self::Canary),
            "dev" | "development" => Ok(Self::Dev),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Feature {
    NativeBedrock,
    OpenAiResponsesStreaming,
    CanaryProbe,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FeatureSpec {
    pub name: &'static str,
    pub available_in: ReleaseChannel,
    pub default_enabled_in: Option<ReleaseChannel>,
}

impl Feature {
    pub fn spec(self) -> FeatureSpec {
        match self {
            Feature::NativeBedrock => FeatureSpec {
                name: "native_bedrock",
                available_in: ReleaseChannel::Stable,
                default_enabled_in: Some(ReleaseChannel::Stable),
            },
            Feature::OpenAiResponsesStreaming => FeatureSpec {
                name: "openai_responses_streaming",
                available_in: ReleaseChannel::Stable,
                default_enabled_in: Some(ReleaseChannel::Stable),
            },
            Feature::CanaryProbe => FeatureSpec {
                name: "canary_probe",
                available_in: ReleaseChannel::Canary,
                default_enabled_in: None,
            },
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Rollout {
    pub channel: ReleaseChannel,
    pub enabled: BTreeSet<String>,
    pub disabled: BTreeSet<String>,
    pub unsafe_allow_unstable: bool,
}

impl Default for Rollout {
    fn default() -> Self {
        Self {
            channel: ReleaseChannel::Stable,
            enabled: BTreeSet::new(),
            disabled: BTreeSet::new(),
            unsafe_allow_unstable: false,
        }
    }
}

impl Rollout {
    pub fn new(
        channel: ReleaseChannel,
        enabled: impl IntoIterator<Item = String>,
        disabled: impl IntoIterator<Item = String>,
        unsafe_allow_unstable: bool,
    ) -> Self {
        Self {
            channel,
            enabled: enabled.into_iter().map(normalize_feature_name).collect(),
            disabled: disabled.into_iter().map(normalize_feature_name).collect(),
            unsafe_allow_unstable,
        }
    }

    pub fn from_parts(
        channel: &str,
        enabled: &str,
        disabled: &str,
        unsafe_allow_unstable: bool,
    ) -> Self {
        Self::new(
            ReleaseChannel::from_str(channel).unwrap_or_default(),
            split_feature_names(enabled),
            split_feature_names(disabled),
            unsafe_allow_unstable,
        )
    }

    pub fn is_available(&self, feature: Feature) -> bool {
        let spec = feature.spec();
        self.channel.allows(spec.available_in) || self.unsafe_allow_unstable
    }

    pub fn is_enabled(&self, feature: Feature, explicit: bool) -> bool {
        let spec = feature.spec();
        if self.disabled.contains(spec.name) {
            return false;
        }
        if !self.is_available(feature) {
            return false;
        }
        if self.enabled.contains(spec.name) || explicit {
            return true;
        }
        spec.default_enabled_in
            .map(|channel| self.channel.allows(channel))
            .unwrap_or(false)
    }
}

pub fn split_feature_names(raw: &str) -> Vec<String> {
    raw.replace(';', ",")
        .split(',')
        .filter_map(|part| {
            let normalized = normalize_feature_name(part);
            (!normalized.is_empty()).then_some(normalized)
        })
        .collect()
}

pub fn normalize_feature_name(raw: impl AsRef<str>) -> String {
    raw.as_ref().trim().to_ascii_lowercase().replace('-', "_")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stable_channel_blocks_canary_feature_even_when_requested() {
        let rollout = Rollout::from_parts("stable", "canary_probe", "", false);

        assert!(!rollout.is_enabled(Feature::CanaryProbe, false));
        assert!(!rollout.is_enabled(Feature::CanaryProbe, true));
    }

    #[test]
    fn canary_channel_allows_explicit_canary_feature() {
        let rollout = Rollout::from_parts("canary", "", "", false);

        assert!(rollout.is_enabled(Feature::CanaryProbe, true));
    }

    #[test]
    fn unsafe_override_allows_unstable_feature_for_break_glass_only() {
        let rollout = Rollout::from_parts("stable", "canary_probe", "", true);

        assert!(rollout.is_enabled(Feature::CanaryProbe, false));
    }

    #[test]
    fn disable_list_wins_over_defaults() {
        let rollout = Rollout::from_parts("stable", "", "native-bedrock", false);

        assert!(!rollout.is_enabled(Feature::NativeBedrock, false));
    }
}
